tests/end2end/experiments/training_plans/mnist_pytorch_training_plan.py

```diff
@@ -60,3 +60,60 @@ def training_step(self, data, target):
         output = self.model().forward(data)
         loss   = torch.nn.functional.nll_loss(output, target)
         return loss
+
+
+# Here we define the model to be used.
+# You can use any class name (here 'Net')
+class BigModelMyTrainingPlan(TorchTrainingPlan):
+
+    # Defines and return model
+    def init_model(self, model_args):
+        return self.Net(model_args = model_args)
+
+    # Defines and return optimizer
+    def init_optimizer(self, optimizer_args):
+        return torch.optim.Adam(self.model().parameters(), lr = optimizer_args["lr"])
+
+    # Declares and return dependencies
+    def init_dependencies(self):
+        deps = ["from torchvision import datasets, transforms"]
+        return deps
+
+    class Net(nn.Module):
+        def __init__(self, model_args):
+            super().__init__()
+            self.conv1 = nn.Conv2d(1, 32, 3, 1)
+            self.conv2 = nn.Conv2d(32, 64, 3, 1)
+            self.dropout1 = nn.Dropout(0.25)
+            self.dropout2 = nn.Dropout(0.5)
+            self.fc1 = nn.Linear(9216, 12800)
+            self.fc2 = nn.Linear(12800, 10)
+
+        def forward(self, x):
+            x = self.conv1(x)
+            x = F.relu(x)
+            x = self.conv2(x)
+            x = F.relu(x)
+            x = F.max_pool2d(x, 2)
+            x = self.dropout1(x)
+            x = torch.flatten(x, 1)
+            x = self.fc1(x)
+            x = F.relu(x)
+            x = self.dropout2(x)
+            x = self.fc2(x)
+
+            output = F.log_softmax(x, dim=1)
+            return output
+
```

Comment:
In my opinion such model is not that big (as big as what we have in tutorials). Only number of neurons in the linear layer has increased. You may add more layers, or use existing models such as `DenseNet`, `Unet`, ..., (some are provided in `pytorch`) as long as the CI slaves can handle it of course.
Also it is usually  good practice to use multiple of `2` as  number of neurons

Reply:
Hi @ybouilla, 

Thank you for the suggestion but the purpose of this example is not creating a real neural network to do training. The purpose is to test mostly gRPC layer to see how it behaves while exchanging big model files.  If you see in the example it applies dry run. 

However, I think the model size is quite big around 600MB.

Reply:
Yes this part I understood! But I think we should use bigger models holding a higher number of neurons.

For example, with the model `vit_h_14` from `torchvision.model`, you can get a model of size 2GB.

I think we should use these kind of models size for stress tests, provided CI slaves can handle it

Reply:
Ok I see your point but could you please tell what do you want to test by providing `vit_h_14` in order to increase number of neurons?

Reply:
I thought we wanted to test that gRPC can handle exchange of heavy files over several rounds, is that what you were testing?

Reply:
Yes exactly, and the model has more than `12800` neurons in the layer FC1 and FC2 so total 25600 neurons. Once it is exported through `torch.save` it is around 500MB, and I think it is enough to test gRPC, because gRPC streams the large message by dividing into 4MB chuncks.

Reply:
Hi @ybouilla, 

All endurance tests are passing now. If it is okay for you I am going to squash the commits, push and merge.

Reply:
> Yes exactly, and the model has more than 12800 neurons in the layer FC1 and FC2 so total 25600 neurons. Once it is exported through torch.save it is around 500MB, and I think it is enough to test gRPC, because gRPC streams the large message by dividing into 4MB chuncks.

Sorry, I am quite confused...

- You want to do a stress test on a high number of Rounds, correct? Ok then for that I think we can just increase the number of `Rounds` to something close to `100` (if i remember correctly, there was a failure on Round 70~80 . And we should rename the test called `e2e_big_neural_network` by something more explicit `e2e_high_number_of_rounds_torch`
- Otherwise, there is not too much difference between the model size of `e2e_mnist_pytorch_training,py` and this model. (they are both around `500MB)`. So you are basically doing the same test several time

Reply:
Besides `12800` is not a valid number of neurons, for there is no valid reason not to use a power of `2`

Reply:
> Please see e2e_increased_number_of_nodes.py and read module doc string


You are doing the test only with  scikit-learn, this is why i was wondering if you wanted to do it on pytorch models, hence the confusion. Maybe another test on pytorch may prove useful


> Are you sure? because I tested classical MNIST one is around 5MB


Sorry, you are right, it depends on how you save the model but you can reach 5MB with classical MNIST model. So ok for this test (even if we could try with bigger models)

> e2e_big_neural_network.py -> to test big model exchange through gRPC. This test DOES NOT test convergence or cares about whether number of neurons are valid or not. It tests gRPC large message exchange.

I never talked about caring for convergence. When I meant by number of neurons are valid, it does mean that the model will run in a more optimized way. As a rule of thumb, it is recommended  to use a power of `2` for neurons (for hardware specific and memory reasons). At the very moment you have some training operation done (even if `dry_run=True`), i think it can be good to optimize model training on hardware. But this is merely a recommendation, so do as you want, the ore important part is that tests are passing

Reply:
> 4.1 Linear quantization
For reasons that will become clear shortly, we used 8-bit quantization to convert activations into
unsigned char, intermediate layer weights into signed char, with the exception of biases
that are encoded as 32-bit int. One exception is the input layer, which remains floating-point, to
better accommodate the potentially larger dynamic ranges of inputs that are not probabilities. In
our particular use-case, the input layer is much smaller (440x2000) than any subsequent layer and
doesn’t weigh strongly on the overall speed. One of the important benefits of quantizing the majority
of the network down to 8 bits is that the total memory footprint of the network consequently shrinks
by between 3× and 4×.

In https://static.googleusercontent.com/media/research.google.com/en//pubs/archive/37631.pdf

Reply:
I used sklearn because it is smaller than neural networks in terms of model size. We can do more round such as 200. If I use MNIST (5MB) 5 x 200 = 1GB there are 10 nodes 10GB disk space in the slave, where we have 40GB free space so it didn't want to risk it.  The goal of this test is to run more rounds using 10 nodes to test all 10 nodes are orchestrated well through researcher.

Reply:
If you want to add more endurance test please feel free to add.

Reply:
> In https://static.googleusercontent.com/media/research.google.com/en//pubs/archive/37631.pdf

Thank you for this useful document and explaining me again that number of neuron should be power of 2. However, I don't think that it will make a difference for dry run.

Reply:
I pushed a new commit to use valid number of neurons.